subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/SetupLoggingActionExecuter.java

```diff
@@ -30,26 +33,40 @@
 public class SetupLoggingActionExecuter implements BuildExecuter {
     private final BuildExecuter delegate;
     private final LoggingManagerInternal loggingManager;
+    private final ParallelExecutionManager parallelExecutionManager;
 
-    public SetupLoggingActionExecuter(BuildExecuter delegate, LoggingManagerInternal loggingManager) {
+    public SetupLoggingActionExecuter(BuildExecuter delegate, LoggingManagerInternal loggingManager, ParallelExecutionManager parallelExecutionManager) {
         this.delegate = delegate;
         this.loggingManager = loggingManager;
+        this.parallelExecutionManager = parallelExecutionManager;
     }
 
     @Override
     public Object execute(BuildAction action, BuildRequestContext requestContext, BuildActionParameters actionParameters, ServiceRegistry contextServices) {
         StartParameter startParameter = action.getStartParameter();
         loggingManager.setLevelInternal(startParameter.getLogLevel());
-        if (startParameter.isParallelProjectExecutionEnabled()) {
-            loggingManager.setMaxWorkerCount(startParameter.getMaxWorkerCount());
-        } else {
-            loggingManager.setMaxWorkerCount(1);
-        }
+        setupLogging(startParameter);
+        ParallelismConfigurationListener listener = new ParallelismConfigurationListener() {
+            @Override
+            public void onConfigurationChange(ParallelismConfiguration parallelismConfiguration) {
+                setupLogging(parallelismConfiguration);
+            }
+        };
+        parallelExecutionManager.addListener(listener);
         loggingManager.start();
         try {
             return delegate.execute(action, requestContext, actionParameters, contextServices);
         } finally {
             loggingManager.stop();
+            parallelExecutionManager.removeListener(listener);
+        }
+    }
+
+    private void setupLogging(ParallelismConfiguration parallelismConfiguration) {
+        if (parallelismConfiguration.isParallelProjectExecutionEnabled()) {
```

Comment:
:o: Shouldn't `maxWorkers` always be `1` when `parallelProjectExecutionEnabled` is `false`? If yes, then this could be simplified.